paddle/fluid/framework/ir/memory_optimize_pass/share_mem_opt_info_to_subgraph_pass.cc

```diff
@@ -0,0 +1,148 @@
+  for (auto* var : compute_op->Outputs()) {
+    if (!var->Node()->IsCtrlVar()) {
+      continue;
+    }
+    for (auto* op : var->PendingOps()) {
+      auto* eager_deletion_op =
+          dynamic_cast<details::EagerDeletionOpHandle*>(op);
+      if (eager_deletion_op) {
+        return eager_deletion_op;
+      }
+    }
+  }
+  return nullptr;
+}
+
+static void ShareVarInfoToCinnLaunch(
+    const MemOptVarInfoMapList& var_infos,
+    details::ComputationOpHandle* cinn_launch_op) {
+  auto* followed_eager_deletion_op =
+      FindFollowedEagerDeletionOp(cinn_launch_op);
+  if (!followed_eager_deletion_op) {
+    VLOG(4) << "No eager_deletion op found after this cinn_launch op";
+    return;
+  }
+  auto vars_to_delete = followed_eager_deletion_op->VarsToDelete();
+  if (vars_to_delete.empty()) {
+    VLOG(4) << "No var to be deleted after this cinn_launch op";
+    return;
+  }
+  VLOG(4) << "Variables would be deleted by eager_deletion_op"
+          << " following the cinn_launch:"
+          << paddle::string::join_strings(vars_to_delete, ',');
+
+  const auto& subgraph = paddle2cinn::CinnCompiler::GetInstance()->FindGraph(
+      cinn_launch_op->GetOp()->Attr<std::string>(operators::kCompilationKey));
+  auto& varinfo_from_maingraph =
+      subgraph.Get<Name2VarInfoMap>(paddle2cinn::kMemOptVarInfoFromMainGraph);
+  const auto& cur_place_var_infos = var_infos.at(cinn_launch_op->GetScopeIdx());
+
+  // collect all MemOptVarInfos of external variables
+  // that would be eager deleted after the cinn_launch subgraph executed,
+  // and store them as attribute of the subgraph
+  for (const auto& var_name : vars_to_delete) {
+    auto it = cur_place_var_infos.find(var_name);
+    PADDLE_ENFORCE_NE(it, cur_place_var_infos.end(),
+                      platform::errors::NotFound(
+                          "MemOptVarInfo of var[%s] not found", var_name));
+    varinfo_from_maingraph.emplace(var_name, it->second);
+  }
+}
+
+static void TakeVarInfoFromMainGraph(
+    const Name2VarInfoMap& parent_var_infos,
+    const MemOptVarInfoMapList& var_infos,
+    details::EagerDeletionOpHandle* eager_deletion_op) {
+  const auto& cur_place_var_infos =
+      var_infos.at(eager_deletion_op->GetScopeIdx());
+  for (auto&& var_name : eager_deletion_op->VarsToDelete()) {
+    auto cur_it = cur_place_var_infos.find(var_name);
+    PADDLE_ENFORCE_NE(cur_it, cur_place_var_infos.end(),
+                      platform::errors::NotFound(
+                          "MemOptVarInfo of var[%s] not found", var_name));
+    auto parent_it = parent_var_infos.find(var_name);
+    if (parent_it != parent_var_infos.end()) {
+      VLOG(4) << "MemOptVarInfo of var[" << var_name << "] set parent holder";
+      cur_it->second->SetParentHolder(parent_it->second);
+    }
+  }
+}
+
+// This pass will be applied on both the main graph and all subgraphs,
+// and it distinguishs them according to whether the graph has the
+// kMemOptVarInfoFromMainGraph attribute or not.
```

Comment:
If the pass is only used in CINN, should we use another name? (Distinguish from cinn subgraph and non-cinn sub_graph)

Reply:
that sounds more reasonable and i change to  `share_varinfo_into_cinn_pass` instead.